cmake: west: Detect whether inside an installation

carlescufi opened the pull request:
If one invokes cmake with west in the PATH but not inside a west
installation (i.e. in a monorepo setup), west will try to list the
zephyr modules issuing an error message. Using the --version option,
detect if west is being invoked from within an installation and disable
it in the negative case to avoid such errors.

Fixes #14177

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

SebastianBoe commented on cmake/host-tools.cmake:
```diff
@@ -7,6 +7,19 @@ find_program(
   unset(WEST)
+else()
+  execute_process(
+  COMMAND
```
indentation

carlescufi replied:
Fixed

SebastianBoe commented on cmake/host-tools.cmake:
```diff
@@ -7,6 +7,19 @@ find_program(
 if(${WEST} STREQUAL WEST-NOTFOUND)
   unset(WEST)
+else()
+  execute_process(
```
A comment above execute process here would be very helpful. E.g.

"When west is found check if we are inside a west installation, and if so mark west as not found because X."

SebastianBoe replied:
It kind of sounds like we would want to have two variables, WEST, and IS_INSIDE_WEST_INSTALLATION. But I'm not familiar enough with this domain to be sure.

carlescufi replied:
Fixed. I don't think we need the two variables. west is essentially useless when not in an installation. It's like Git outside of a repository.

SebastianBoe replied:
Doesn't the build system use WEST for flashing?

E.g. won't this affect zephyr/cmake/flash/CMakeLists.txt ?

tejlmand replied:
> E.g. won't this affect zephyr/cmake/flash/CMakeLists.txt ?

that is already broken, and this PR doesn't change that.
It is actually a very valid question if that should be fixed as well, but that would need a fix in `west`, and might not be so trivial, as `west flash` is an extension command, so it would require at least (as project is mono-repo is managed without `west`):
1) `west` repo cloned somewhere, and path known to it know, e.g. `WEST_DIR`
2) Ability in `west` to load extension commands at different location, as it might not be located together with a `manifest` project

SebastianBoe replied:
> that is already broken

My mistake, I was not aware. Please do fix/remove any broken code.

tejlmand replied:
> >    that is already broken

> My mistake, I was not aware. Please do fix/remove any broken code.

Me neither, I only discovered this, because you dragged my attention to this with your comment.
And could see this must have been broken for some time, unfortunately :(

tejlmand replied:
actually, flashing should still work if following this guide: https://docs.zephyrproject.org/latest/guides/west/without-west.html

tejlmand commented:
@carlescufi Regarding this:
>  Using the --version option,
detect if west is being invoked from within an installation and disable
it in the negative case to avoid such errors.

This PR actually doesn't prevent out-of-tree.
If `ZEPHYR_BASE` is set, and pointing to a zephyr inside an installation, `west --version` will give that version.

I think this is correct behavior, but just wanted to make you aware that out-of-tree is possible, as long as `ZEPHYR_BASE` is set and points to something valid.

So correct, this PR is, avoid `cmake: west: Allow mono-repo to build when west is installed`

```
$ pwd
/tmp
$ echo $ZEPHYR_BASE
/projects/<path>/zephyr
$ west --version
West bootstrapper version: v0.5.4 (...)
West repository version: v0.5.4-12-gcc5a05f (/projects/<path>/.west/west)
```

tejlmand committed:
cmake: west: Allow mono-repo to build when west is installed

If one invokes cmake with west in the PATH but not inside a west
installation (i.e. in a monorepo setup), west will try to list the
zephyr modules issuing an error message.

Test if west list succeeds before using result for module testing.

Fixes #14177

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

tejlmand commented on cmake/host-tools.cmake:
```diff
@@ -7,6 +7,24 @@ find_program(
+  execute_process(
+    COMMAND
+    ${WEST} --version
+    OUTPUT_VARIABLE west_version_output
```
The alternative approach can be found here: https://github.com/tejlmand/zephyr/commit/0fd80529edceb4122ccafb8d36e63e607b6943e6

carlescufi replied:
>I would like to suggest an alternative, which is to simply do the check in west list and if that fails, then there are no modules to import as that indicates an mono-repo installation.

We discussed this offline yesterday with @mbolivar. The issue is that `west list` can fail for other reasons, such as a malformed manifest. That makes it then confusing for the user if west is indeed correctly inside an installation but the manifest has some mistaken in it.

tejlmand replied:
I still think changing the approach makes sense.

In future `west` may be valid for other commands, even if `west list` can fail in a mono-repo install, example can be `west config <name>` that can fetch general `west` settings, similar to `git config` can be used outside a git repo. So completely disabling `west` may not be that desired, as `west` is indeed installed.

But I acknowledge that `west list` can fail for other reasons, and when that happens the message should still be printed. Thus I updated the proposal here: https://github.com/tejlmand/zephyr/commit/b92ec0d4dae6c6210932a13e63083c5094b914e8 so that in case the failure is not because we are outside a `west` managed project, then the error is still printed.

tejlmand reviewed (commented):
A couple of comments and an alternative solution to considder.

SebastianBoe reviewed (approved):
CMake-related changes are OK. But I am not able to review the west-integration specifics.
